aws-service-discovery/src/main/java/io/micronaut/discovery/aws/servicediscovery/registration/AwsServiceDiscoveryRegistrationClient.java

```diff
@@ -0,0 +1,290 @@
+    private Service discoveryService;
+    private final Executor executorService;
+    private final ServiceDiscoveryClient serviceDiscoveryClient;
+    private final ServiceDiscoveryAsyncClient serviceDiscoveryAsyncClient;
+
+    /**
+     * Constructor for setup.
+     *
+     * @param environment                                  current environment
+     * @param awsServiceDiscoveryRegistrationConfiguration config for auto registration
+     * @param amazonComputeInstanceMetadataResolver        resolver for aws compute metdata
+     * @param executorService                              this is for executing the thread to monitor the register operation for completion
+     * @param serviceDiscoveryClient                       the serviceDiscoveryClient
+     * @param serviceDiscoveryAsyncClient                  the serviceDiscoveryAsyncClient
+     */
+    protected AwsServiceDiscoveryRegistrationClient(
+            Environment environment,
+            AwsServiceDiscoveryRegistrationConfiguration awsServiceDiscoveryRegistrationConfiguration,
+            AmazonComputeInstanceMetadataResolver amazonComputeInstanceMetadataResolver,
+            @Named(TaskExecutors.IO) Executor executorService,
+            ServiceDiscoveryClient serviceDiscoveryClient,
+            ServiceDiscoveryAsyncClient serviceDiscoveryAsyncClient) {
+        super(awsServiceDiscoveryRegistrationConfiguration);
+        this.environment = environment;
+        this.route53AutoRegistrationConfiguration = awsServiceDiscoveryRegistrationConfiguration;
+        this.serviceDiscoveryClient = serviceDiscoveryClient;
+        this.serviceDiscoveryAsyncClient = serviceDiscoveryAsyncClient;
+        this.amazonComputeInstanceMetadataResolver = amazonComputeInstanceMetadataResolver;
+        this.executorService = executorService;
+    }
+
+    /**
+     * If custom health check is enabled, this sends a heartbeat to it.
+     * In most cases aws monitoring works off polling an application's endpoint
+     *
+     * @param instance The instance of the service
+     * @param status   The {@link HealthStatus}
+     */
+    @Override
+    protected void pulsate(ServiceInstance instance, HealthStatus status) {
+        // this only work if you create a health status check when you register it
+        // we can't really pulsate anywhere because amazon health checks work inverse from this UNLESS you have a custom health check
+        Optional<String> opt = instance.getInstanceId();
+        if (!opt.isPresent()) {
+            // try the metadata
+            if (instance.getMetadata().contains("instanceId")) {
+                opt = Optional.of(instance.getMetadata().asMap().get("instanceId"));
+            } else {
+                LOG.error("Cannot determine the instance ID. Are you sure you are running on AWS EC2?");
+            }
+        }
+
+        opt.ifPresent(instanceId -> {
+            if (discoveryService != null && discoveryService.healthCheckConfig() != null) {
```

Comment:
Any idea what is `discoveryService` supposed to be? It's only set in tests but it looks like it should be provided from somewhere.

Reply:
good question

Reply:
I did more digging and it used to be set during the registration:
```
discoveryService = getService(route53AutoRegistrationConfiguration.getAwsServiceId());
```
it was removed in the initial version of this class.